src/vs/workbench/api/browser/mainThreadCommands.ts

```diff
@@ -78,9 +78,21 @@ export class MainThreadCommands implements MainThreadCommandsShape {
 		return this._commandService.executeCommand<T>(id, ...args);
 	}
 
+	$onDidExecuteCommand() {
+		return this._commandService.onDidExecuteCommand((command) => this.handleExecuteCommand(command));
+	}
+
+	$disposeListeners() {
+		return this._commandService.disposeListeners();
```

Comment:
Added `$disposeListeners` to call the `dispose` on the main thread event emitter when there are no more listeners subscribed on the extension side.  Seem legit?

Reply:
No, this is not how it works. Your implementation of `CommandService#disposeListeners` removes all listeners and kills the `CommandService#onDidExecuteCommand` event. 

* `ICommandService#onDidExecuteCommand` is an even with potentially *many* listeners.
* `MainThreadCommands` can be a (1**!**) listener for such event. Its implementation forwards the event data to the extension host process
* The extension host process (`extHostCommands`) receives the event data and all it does is firing another event. That event (`Emitter`) is exposed as API and manages its own set of listeners. That means when an "API-listener" stops listening nothing needs to change in the `ICommandService`

Reply:
Got it.  To be clear, should I still add/kill the listener on first add/last remove on the `Emitter` in `extHostCommands`, but only get rid of the listener in `MainThreadCommands`?  

```
this._onDidExecuteCommand = new Emitter<ICommandEvent>({
	onFirstListenerDidAdd: this._proxy.$onDidExecuteCommand,
	onLastListenerRemove: this._proxy.$disposeListener,
});
```

I assume that's preferable to having the listener in `MainThreadCommands` always forward events despite no listeners in the extension side `Emitter`, but just want to run it by you first.  Thanks for the assistance, btw.

Reply:
Just updated the code to add/dispose of the single listener

Reply:
Yeah, think of it (and tackle it) as two problems. 

1) to implement the `onDidExecuteCommand` on the extension host side you need the forwarding so that you can drive an `Emitter` there
2) now when no-one listens on the extension host side you still forward all those events for nothing, wasting IPC messages/resource. that's where the added first, removed last listener comes into play.

Reply:
Got it.  Just wanted to confirm.  I pushed up changes to add/remove only the single listener in `MainThreadCommands`.  Should be good to go.